### PR TITLE
Added tests for template generation

### PR DESCRIPTION
**Pull request by santacodes**
# Description 
- Added basic tests for template generation using `pytest-cookies`.
- Added `cookiecutter` as a dependency and setup `___init.py__` for the src layout.

### CODE REVIEW

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Thanks, @santacodes! Looks good to me. Could you please add a minimal CI job to check if the tests are working fine?

File: tests/test_project_generation.py
Comment (arjxn-py) [inline tests/test_project_generation.py:21]:
```suggestion
def test_bake_custom_project(cookies: Cookies, custom_template):
```

Don't want to make the work harder but can we also take care of typing a bit from the start just in case you're familiar with it?

File: tests/test_project_generation.py
Comment (arjxn-py) [inline tests/test_project_generation.py:2]:
```suggestion
import pytest
from pytest_cookies.plugin import Cookies
```

Related to the other typing comment.

**Review by arjxn-py — COMMENTED** [review]
Thanks @santacodes, Good Start I must admit :) 
Some comments below to propose some improvements

File: tests/test_project_generation.py
Comment (santacodes) [reply tests/test_project_generation.py:9]: Sorry for the late reply, I was travelling, I would also mention `tmp_path` returns [py.path](https://py.readthedocs.io/en/latest/path.html) which also quotes 
> Note: The ‘py’ library is in “maintenance mode” and so is not recommended for new projects. Please check out [pathlib](https://docs.python.org/3/library/pathlib.html) or [pathlib2](https://pypi.org/project/pathlib2/) for path operations.

So I am not sure which one shall I proceed with.

**Comment by santacodes** [comment]
> Thanks, @santacodes! Looks good to me. Could you please add a minimal CI job to check if the tests are working fine?

Sorry for the delay @Saransh-cpp, I was travelling. I added CI for the tests, created a nox session for testing the template generation, and tested it on my branch. Could you please review and look at it? Also, I added uv to the nox backend, which seems to be faster.

File: tests/test_project_generation.py
Comment (santacodes) [reply tests/test_project_generation.py:29]: Added them in the latest commit, do they seem fine or would we be better with the usual `try-except` blocks?

File: noxfile.py
Comment (Saransh-cpp) [inline noxfile.py:29]:
```suggestion
    session.install("-e", ".[dev]", silent=False)
```

File: src/pybamm_cookiecutter/__init__.py
Comment (Saransh-cpp) [inline src/pybamm_cookiecutter/__init__.py:14]: pybamm should not be importable through this module
```suggestion
```

File: .github/test_on_push.yml
Comment (Saransh-cpp) [inline .github/test_on_push.yml:5]: Should we also run it with `push: [main]`?

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Thanks, @santacodes! Few minor changes below -

**Comment by Saransh-cpp** [comment]
TODO:
- [ ] Update branch protection rule to include the new CI (status checks must pass before merging)

File: .github/test_on_push.yml
Comment (santacodes) [reply .github/test_on_push.yml:5]: added it in the latest commit

**Comment by santacodes** [comment]
> TODO:
> 
> * [ ]  Update branch protection rule to include the new CI (status checks must pass before merging)

branchrule

I think I can only create a branch rule allowing merges only after status checks when the status checks themselves have been merged to the main branch.

**Comment by Saransh-cpp** [comment]
Hi, don't worry about the rules. I'll set them up.

**Review by Saransh-cpp — APPROVED** [review]
This looks good to me, thanks @santacodes! I'll wait for @agriyakhetarpal and @arjxn-py's reviews.

**Comment by arjxn-py** [comment]
Hi, I'll review this in a couple of hours.

File: tests/test_project_generation.py
Comment (arjxn-py) [inline tests/test_project_generation.py:27]:
```suggestion
    assert result.exit_code == 0, f"Exited with code {result.exit_code}, expected 0"
```
This should be more helpful, correct me if needed :)

File: tests/test_project_generation.py
Comment (arjxn-py) [inline tests/test_project_generation.py:30]:
```suggestion
    assert result.project_path.is_dir(), f"Project directory {result.project_path} not found"
```

File: tests/test_project_generation.py
Comment (arjxn-py) [reply tests/test_project_generation.py:29]: It's perfect, we don't need try-expect blocks for this.

**Review by arjxn-py — COMMENTED** [review]
Looks much better @santacodes 🙂, Some comments below for minor improvements, feel free to ping me again.
Before we merge, we would want that the workflow starts running on this same PR.
Plus, I really appreciate that we're also adapting to `uv` in this. 
xref [PyBaMM#3825](https://github.com/pybamm-team/PyBaMM/issues/3825)

File: .github/test_on_push.yml
Comment (agriyakhetarpal) [inline .github/test_on_push.yml:1]:
```suggestion
name: Test template generation
```

File: .github/test_on_push.yml
Comment (agriyakhetarpal) [inline .github/test_on_push.yml:30]:
```suggestion
        os: [ubuntu-latest, macos-13, macos-14, windows-latest]
```

File: .github/test_on_push.yml
Comment (agriyakhetarpal) [inline .github/test_on_push.yml:35]:
```suggestion
      - name: Checkout pybamm-cookiecutter
```

File: .github/test_on_push.yml
Comment (agriyakhetarpal) [inline .github/test_on_push.yml:48]: Question: is it possible to cache `uv` through this action? I haven't explored it. We should cache either `pip` or `uv`, since both of them store cached wheels in difference places, IIRC.

File: pyproject.toml
Comment (agriyakhetarpal) [inline pyproject.toml:35]: I would suggest changing `nox` in this table to `nox[uv]`, otherwise it will always use `virtualenv`.

File: pyproject.toml
Comment (agriyakhetarpal) [inline pyproject.toml:39]:
```suggestion
  "nox[uv]",
```

File: noxfile.py
Comment (agriyakhetarpal) [inline noxfile.py:25]:
```suggestion
    """Run the tests for testing template generation"""
```

File: tests/test_project_generation.py
Comment (agriyakhetarpal) [reply tests/test_project_generation.py:9]: I think `tmp_path` should be better, then. It's easier to work with `pathlib.Path`

**Comment by agriyakhetarpal** [comment]
P.S. I'm sorry for the late review! I have been a bit out of touch, plus I noticed that Saransh and Arjun had already put in their reviews. I hope this is helpful.

**Review by arjxn-py — APPROVED** [review]
Thanks @santacodes, the tests look perfect to me now.

**Comment by arjxn-py** [comment]
Before we merge, I'd just want to confirm if we can safely resolve the conversation about `uv` caching plus can you maybe link the CI run on the latest commit from your fork here?

File: .github/test_on_push.yml
Comment (agriyakhetarpal) [reply .github/test_on_push.yml:48]: Yes, makes sense – let's look at `uv`'s caching later during the project. As an aside, I did a quick run for `uv` in my PyBaMM repository locally and it works quite well now, saving up a minute of time and more.

File: .github/test_on_push.yml
Comment (agriyakhetarpal) [inline .github/test_on_push.yml:51]:
```suggestion
        run: uv pip install nox[uv]
```

**Review by agriyakhetarpal — DISMISSED** [review]
This looks good to me as long as the tests pass – resolved the previous conversation. Thanks for adding this, @santacodes!

xrefs for `uv`'s caching to look at later: https://github.com/astral-sh/uv/issues/2231, https://github.com/yezz123/setup-uv/issues/25, https://github.com/actions/setup-python/issues/822, https://github.com/actions/setup-python/pull/818

**Review by agriyakhetarpal — APPROVED** [review]

**Comment by agriyakhetarpal** [comment]
(I think we should disable the setting that dismisses reviews on further commits)

**Comment by arjxn-py** [comment]
I'm merging this as @santacodes needs to start working on entrypoints as well.

**Comment by arjxn-py** [comment]
> (I think we should disable the setting that dismisses reviews on further commits)

Yes, I was going to mention that reviews were getting dismissed automatically

**Pull request merged**
